juju/juju
# [JUJU-3502] Slow query logging

To better understand how queries are running in a controller, it is sometimes helpful
to see the slow queries impacting the system at large. Cases like missing indexes or
incorrect queries can cause the controller to run slowly, impacting the performance of
a model.

To somewhat mirror what we have with MongoDB, where they have a top query 
analyzer, we can implement something similar here. Ideally used in conjunction with
the db accessor metrics https://github.com/juju/juju/pull/15295, we can better identify what queries are running slowly.

The dqlite library provides a way to trace how long a query took. We hook into that
system to analyze the tracing log before determining if a query is slow or not.

The feature is disabled by default, as we're using a slightly expensive runtime function
call `runtime.Stack()` to pinpoint exactly the cause of the slow query. Tuning the slow
query threshold allows the fine-tuning of what queries should be surfaced. Also 
depending on the provider/substrate, machine, and workload changing this to another 
than the default 1 second is allowed.

----

Implementation-wise, we're using tracing logging to match the actual trace string, which
isn't ideal! We can look to making this robust in the future, maybe via context tagging.

For now, the parsing determines if we have a normal message, a slow, or an ignored 
query. Changing the query threshold or enablement is done via a controller config.
Changing the value will cause a restart of the agent(s) if the agent is already running.

A follow-up PR will move the log into a slow query log on the unit. For now, this PR is
already complex enough.

## Checklist


## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju controller-config query-tracing-enabled=true
$ juju controller-config query-tracing-threshold=1ms 
```

You can see the slow log queries in debug log.

```sh
$ juju debug-log -m controller
```

Example:

```sh
2023-04-14 15:49:16 WARNING juju.worker.dbaccessor node.go:251 slow query: 0.005s request exec: "COMMIT"
goroutine 4571 [running]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x65
github.com/juju/juju/database.(*NodeManager).slowQueryLogFunc.func1(0x611c77c?, {0x616dcab, 0x16}, {0xc00119cf00?, 0x2, 0x2})
        /home/simon/go/src/github.com/juju/juju/database/node.go:251 +0x98
github.com/canonical/go-dqlite/driver.(*Conn).ExecContext(0xc000305080, {0x6aab620, 0xc000078080}, {0x611c77c, 0x6}, {0x0?, 0x0?, 0x0?})
        /home/simon/go/pkg/mod/github.com/canonical/go-dqlite@v1.11.9/driver/driver.go:397 +0x24e
github.com/canonical/go-dqlite/driver.(*Tx).Commit(0xc0018e0ec0)
        /home/simon/go/pkg/mod/github.com/canonical/go-dqlite@v1.11.9/driver/driver.go:518 +0x47
database/sql.(*Tx).Commit.func1()
        /usr/local/go/src/database/sql/sql.go:2268 +0x2f
database/sql.withLock({0x6a6b998, 0xc000ea82d0}, 0xc0010c56d0)
        /usr/local/go/src/database/sql/sql.go:3405 +0x8c
database/sql.(*Tx).Commit(0xc000db9600)
        /usr/local/go/src/database/sql/sql.go:2267 +0x10f
github.com/juju/juju/database/txn.(*TransactionRunner).Txn(0xc000472500, {0x6aab5e8?, 0xc00069aa50?}, 0x18?, 0xc000d5d020)
        /home/simon/go/src/github.com/juju/juju/database/txn/transaction.go:123 +0x218
github.com/juju/juju/database.Txn(...)
        /home/simon/go/src/github.com/juju/juju/database/txn.go:25
github.com/juju/juju/worker/dbaccessor.(*trackedDBWorker).TxnNoRetry(0xc000c27380, {0x6aab5e8, 0xc00069aa50}, 0x69462a?)
        /home/simon/go/src/github.com/juju/juju/worker/dbaccessor/tracker.go:141 +0x2f7
github.com/juju/juju/worker/dbaccessor.(*trackedDBWorker).Txn.func1()
        /home/simon/go/src/github.com/juju/juju/worker/dbaccessor/tracker.go:117 +0x29
github.com/juju/retry.Call({0xc0008ad9b0, 0xc00337e1e0, 0x0, 0xfa, 0xf4240, 0x5f5e100, 0x5d21dba00, 0xc0027b4720, {0x6a8ebf0, 0x9d6f1e8}, ...})
        /home/simon/go/pkg/mod/github.com/juju/retry@v1.0.0/retry.go:183 +0xdb
github.com/juju/juju/database/txn.defaultRetryStrategy.func1({0x6aab5e8, 0xc00069aa50}, 0xc0008ad9b0)
        /home/simon/go/src/github.com/juju/juju/database/txn/transaction.go:142 +0x1a5
github.com/juju/juju/database/txn.(*TransactionRunner).Retry(...)
        /home/simon/go/src/github.com/juju/juju/database/txn/transaction.go:134
github.com/juju/juju/database.Retry(...)
        /home/simon/go/src/github.com/juju/juju/database/txn.go:35
github.com/juju/juju/worker/dbaccessor.(*trackedDBWorker).Txn(0xc000c27380?, {0x6aab5e8?, 0xc00069aa50?}, 0xc000d5d020?)
        /home/simon/go/src/github.com/juju/juju/worker/dbaccessor/tracker.go:116 +0xc2
github.com/juju/juju/worker/lease.(*Store).ExtendLease(0xc000480120, {0x6aab5e8, 0xc00069aa50}, {{0x6171cb1, 0x16}, {0xc0004af956, 0x24}, {0xc0011e37c4, 0xa}}, {{0xc00348bbd0, ...}, ...})
        /home/simon/go/src/github.com/juju/juju/worker/lease/store.go:139 +0x135
github.com/juju/juju/worker/lease.(*Manager).handleClaim(0xc0004aa000, {0x6aab5e8, 0xc00069aa50}, {{{0x6171cb1, 0x16}, {0xc0004af956, 0x24}, {0xc0011e37c4, 0xa}}, {0xc00348bbd0, ...}, ...})
        /home/simon/go/src/github.com/juju/juju/worker/lease/manager.go:376 +0x5b8
github.com/juju/juju/worker/lease.(*Manager).retryingClaim(0xc0004aa000, {0x6aab5e8, 0xc00069aa50}, {{{0x6171cb1, 0x16}, {0xc0004af956, 0x24}, {0xc0011e37c4, 0xa}}, {0xc00348bbd0, ...}, ...})
        /home/simon/go/src/github.com/juju/juju/worker/lease/manager.go:275 +0x130
created by github.com/juju/juju/worker/lease.(*Manager).choose
        /home/simon/go/src/github.com/juju/juju/worker/lease/manager.go:206 +0x5b2
```
